File: src/_2026/DynamicProgramming/MaxProfit.java
package _2026.DynamicProgramming;

import java.util.Arrays;

public class MaxProfit {
    public int bruteForceMethod(int[] prices) {
        int max = 0;

        for (int i = 0; i < prices.length; i++) {
            for (int j = i + 1; j < prices.length; j++) {
                int calculated = prices[j] - prices[i];
                if (calculated > max) {
                    max = calculated;
                }
            }
        }

        return max;
    }

    // Tried
    public int dpMethod(int[] prices) {

        int buyPrice = prices[0];
        int profit = 0;

        for (int i = 1; i < prices.length; i++) {
            if(prices[i] < buyPrice) {
                buyPrice = prices[i];
            }else{
                profit = Math.max(profit,  prices[i] - buyPrice);
            }
        }

        return profit;
    }
}

File: src/_2026/Python/Arrays/max_profit.py
def max_profit(prices: list[int]) -> int:
    profit = 0
    buy = prices[0]

    for price in range(len(prices)):
        if prices[price] < buy:
            buy = prices[price]
        elif prices[price] - buy > profit:
            profit = prices[price] - buy

    return profit

print(max_profit([7, 1, 5, 3, 6, 4]))
